File: Leetcode/Leetcode-DSA-course/Stacks-and-Queues/496-next-greater-element-i/next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        stack = []
        map = {}
        for i in range(len(nums2)):
            while stack and nums2[i] > stack[-1]:
                map[stack.pop()] = nums2[i]
            stack.append(nums2[i])
        while stack:
            map[stack.pop()] = -1

        ans = []
        for num in nums1:
            ans.append(map[num])
        return ans

File: Leetcode/Leetcode-DSA-course/Stacks-and-Queues/496-next-greater-element-i/next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        stack<int> stack;
        unordered_map<int, int> map;
        for (int i = 0; i < nums2.size(); i++) {
            while (!stack.empty() && nums2[i] > stack.top()) {
                map[stack.top()] = nums2[i];
                stack.pop();
            }
            stack.push(nums2[i]);
        }
        while (!stack.empty()) {
            map[stack.top()] = -1;
            stack.pop();
        }
        vector<int> ans;
        for (int num : nums1)
            ans.push_back(map[num]);

        return ans;
    }
};